File: MATKIT/TESTS/test_matkit_efficiency.cpp
// this driver is for testing the efficiency of some MATKIT functions

#include <iostream>  // for cout, cerr, endl, etc. (under namespace std)
#include <time.h>    // for time_t, time, clock_t, clock, CLOCKS_PER_SEC, etc.
#include "matkit.h"

using std::cout;
using std::endl;

#ifdef USE_NAMESPACE
using namespace MATKIT;
#endif


int main(int argc, char *argv[]) {
    ///////////////////////////////////////////////////////////////////////////
    //    efficiency test
    ///////////////////////////////////////////////////////////////////////////
    cout << "### efficiency test" << endl << endl;

    // construct a sparse matrix for tests
    mkIndex nr = 20000, nc = 30000;
    int bnd = 50;
    SparseMatrix S(nr, nc);
    for (int j=1; j<=(int)nc; j++) {
        for (int i=Basics::max(j-bnd,1); i<=Basics::min(j+bnd,(int)nr); i++) {
            S(i,j) = 1.0/(1.0+(Real)Basics::abs(i-j));
        }
    }
    cout << "S is of size " << S.Nrows() << "-by-" << S.Ncols() << " with nnz=" << S.Nnz() << endl;
    cout << "||S||_1   = " << S.norm1() << endl;
    cout << "||S||_F   = " << S.normFrobenius() << endl;
    cout << "||S||_inf = " << S.normInfinity() << endl << endl;

    // sparse matrix - vector multiplication
    cout << "v = Vector::ones(" << S.Ncols() << ")" << endl;
    Vector v = Vector::ones(S.Ncols()), w;
    cout << "w=S*v time used (1000 times): ";
    clock_t start = clock();
    for (int i=0; i<1000; i++)
        w = S*v;
    cout << (Real)(clock()-start)/(Real)CLOCKS_PER_SEC << " seconds" << endl;
    cout << "||w||_1   = " << w.norm1() << endl;
    cout << "||w||_2   = " << w.norm2() << endl;
    cout << "||w||_inf = " << w.normInfinity() << endl << endl;

    cout << "v = Vector::ones(" << S.Nrows() << ")" << endl;
    v = Vector::ones(S.Nrows());
    Vector w2;
    cout << "w2=v'*S time used (1000 times): ";
    start = clock();
    for (int i=0; i<1000; i++)
        w2 = v*S;
    cout << (Real)(clock()-start)/(Real)CLOCKS_PER_SEC << " seconds" << endl;
    cout << "||w2||_1   = " << w2.norm1() << endl;
    cout << "||w2||_2   = " << w2.norm2() << endl;
    cout << "||w2||_inf = " << w2.normInfinity() << endl << endl;

    // dense matrix transpose
    mkIndex dim = 10000;
    Matrix M = Matrix::random(dim, dim);
    cout << "time used to transpose a " << dim << "-by-" << dim << " dense matrix: ";
    start = clock();
    M.transpose();
    cout << (Real)(clock()-start)/(Real)CLOCKS_PER_SEC << " seconds" << endl << endl;

    // dense matrix - vector multiplication
    v = Vector::random(dim);
    start = clock();
    cout << "dense matrix - vector multiplication (n=" << dim << ") time used: ";
    v = M*v;
    cout << (Real)(clock()-start)/(Real)CLOCKS_PER_SEC << " seconds" << endl;

    cout << "vector - dense matrix multiplication (n=" << dim << ") time used: ";
    start = clock();
    v = v*M;
    cout << (Real)(clock()-start)/(Real)CLOCKS_PER_SEC << " seconds" << endl;

    SymmetricMatrix M2 = SymmetricMatrix::random(dim);
    cout << "symmetric matrix - vector multiplication (n=" << dim << ") time used: ";
    start = clock();
    v = M2*v;
    cout << (Real)(clock()-start)/(Real)CLOCKS_PER_SEC << " seconds" << endl << endl;

    // matrix - matrix multiplication, matrix - symmetric matrix multiplication
    dim = 2000;
    M = Matrix::random(dim, dim);
    cout << "dense matrix - dense matrix multiplication (n=" << dim << ") time used: ";
    start = clock();
    Matrix N = M*M;
    cout << (Real)(clock()-start)/(Real)CLOCKS_PER_SEC << " seconds" << endl;
    M2 = SymmetricMatrix::random(dim);
    cout << "dense matrix - symmetric matrix multiplication (n=" << dim << ") time used: ";
    start = clock();
    N = M*M2;
    cout << (Real)(clock()-start)/(Real)CLOCKS_PER_SEC << " seconds" << endl;
    cout << "symmetric matrix - dense matrix multiplication (n=" << dim << ") time used: ";
    start = clock();
    N = M2*M;
    cout << (Real)(clock()-start)/(Real)CLOCKS_PER_SEC << " seconds" << endl;

    return 0;
}

File: MATKIT/TESTS/test_matkit_efficiency.m
% This m-file performs similar operations to those in
% "test_matkit_efficiency.cpp", which is for testing the efficiency of some
% MATKIT functions.


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%    efficiency test
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
fprintf('### efficiency test\n\n');

% sparse matrix - vector multiplication
nr = 20000;
nc = 30000;
bnd = 50;
S = sparse(nr, nc);
for j = 1:nc
    for i = max(1,j-bnd):min(j+bnd,nr)
        S(i,j) = 1/(1.0+abs(i-j));
    end
end
fprintf('S is of size %d-by-%d with nnz = %d\n', size(S,1), size(S,2), nnz(S));
fprintf('||S||_1   = %g\n', norm(S,1));
fprintf('||S||_F   = %g\n', norm(S,'fro'));
fprintf('||S||_inf = %g\n\n', norm(S,inf));

% sparse matrix - vector multiplication
fprintf('v = ones(%d, 1);\n', nr);
v = ones(nc, 1);
fprintf('w=S*v time used (1000 times): ');
t = cputime;
for i = 1:1000
    w = S*v;
end
fprintf('%g seconds\n', cputime-t);
fprintf('||w||_1   = %g\n', norm(w,1));
fprintf('||w||_2   = %g\n', norm(w,2));
fprintf('||w||_inf = %g\n\n', norm(w,inf));

% vector - sparse matrix multiplication
fprintf('v2 = ones(1, %d);\n', nc);
v2 = ones(1, nr);
fprintf('w2=v2*S time used (1000 times): ');
t = cputime;
for i = 1:1000
    w2 = v2*S;
end
fprintf('%g seconds\n', cputime-t);
fprintf('||w2||_1   = %g\n', norm(w2,1));
fprintf('||w2||_2   = %g\n', norm(w2,2));
fprintf('||w2||_inf = %g\n\n', norm(w2,inf));

% dense matrix transpose
dim = 10000;
M = rand(dim, dim);
fprintf('time used to transpose a %d-by-%d matrix: ', dim, dim);
t = cputime;
transpose(M);
fprintf('%g seconds\n', cputime-t);

% dense matrix - vector multiplication
v = rand(dim, 1);
fprintf('dense matrix - vector multiplication (n=%d) time used: ', dim);
t = cputime;
w = M*v;
fprintf('%g seconds\n', cputime-t);

v = rand(1, dim);
fprintf('vector - dense matrix multiplication (n=%d) time used: ', dim);
t = cputime;
w = v*M;
fprintf('%g seconds\n', cputime-t);

% dense matrix - dense matrix multiplication
dim = 2000;
M = rand(dim);
fprintf('dense matrix - dense matrix multiplication (n=%d) time used: ', dim);
t = cputime;
N = M*M;
fprintf('%g seconds\n', cputime-t);
